src/D2L.CodeStyle.Analyzers/Async/Generator/FileCollector.cs

```diff
@@ -0,0 +1,304 @@
+﻿using System.Collections.Immutable;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace D2L.CodeStyle.Analyzers.Async.Generator;
+
+internal partial class SyncGenerator {
+	/// <summary>
+	/// Creates a source file containing generated methods and the necessary
+	/// namespaces, using directives, and type declarations for them to live in
+	/// (but only the necessary ones).
+	/// </summary>
+	internal sealed class FileCollector {
+		private static readonly SyntaxToken NoneToken
+			= SyntaxFactory.Token( SyntaxKind.None );
+
+		private static readonly SyntaxTriviaList EmptyTrivia
+			= SyntaxFactory.TriviaList();
+
+		private readonly CompilationUnitSyntax m_root;
+
+		// This is a mutable dictionary because we will remove from it as we go
+		// and check that it is empty at the end, just to be safe.
+		private readonly Dictionary<SyntaxNode, IEnumerable<string>> m_methods;
+
+		/// <summary>
+		/// Wraps the sort of syntax that we will append as part of a preamble, to
+		/// defer calls to ToFullString()
+		/// </summary>
+		private readonly struct PieceOfSyntax {
+			public object Data { get; init; }
+
+			public static implicit operator PieceOfSyntax( SyntaxNode node)
+				=> new() { Data = node };
+
+			public static implicit operator PieceOfSyntax( SyntaxToken token )
+				=> new() { Data = token };
+
+			public static implicit operator PieceOfSyntax( SyntaxList<UsingDirectiveSyntax> list )
+				=> new() { Data = list };
+
+			public override string ToString() => Data switch {
+				SyntaxNode n => n.ToFullString(),
+				SyntaxToken t => t.ToFullString(),
+				SyntaxList<UsingDirectiveSyntax> nl => nl.ToFullString(),
+				_ => throw new BugException( "invalid piece of syntax " + Data.GetType().Name )
+			};

```

Comment:
See the other thread for context. `SyntaxList<UsingDirectiveSyntax>` may seem really specific, but I don't expect this list of types to grow over time -- this functionality should be relatively "done".